source/opt/ir_builder.h

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#ifndef LIBSPIRV_OPT_SPVBUILDER_H_
+#define LIBSPIRV_OPT_SPVBUILDER_H_
+
+#include <memory>
+
+#include "basic_block.h"
+#include "instruction.h"
+#include "module.h"
+#include "spirv-tools/libspirv.h"
+
+namespace spvtools {
+namespace opt {
+namespace ir {
+
+// Builder class for constructing SPIR-V in-memory IR representation. Methods in
+// this class are designed to work with the interface for spvBinaryParse() in
+// libspirv.h so that we can leverage the syntax checks implemented behind it.
+//
+// The user is expected to call SetModuleHeader() to fill in the module's
+// header, and then AddInstruction() for each decoded instruction, and finally
+// Link() to linking everything together.
+class IrBuilder {
```

Comment:
If this is all this class is going to do, then it's better to call it IRLoader.
Normally, an IR Builder (like in LLVM or in Glslang) is much more helpful in constructing a module.  E.g. being able to set the build point at a random spot in the module, keeping type tables, etc.